### ISSUE TITLE
Caching should be optional

### ISSUE THREAD

**Issue by ryanwarsaw** [SunshineAPI/WebAPI#45]
Currently your software caches things, but there are no instructions on how to turn this feature off, if you can even turn it off at all? Caching should be an optional feature that people can enable or disable freely.

**Comment by gilrosenthal** [comment]
@ryanw-se  Why? The redis server is already up for authentication, and it is not like it has a huge load on the server.
Also, the main advantage of sunshine is that it does not DDOS OCN, which we need if we are to stay alive. You can edit the caching times in config.js, and you can set it to smaller times or even 0 seconds if you really want. It just is not worth it

**Comment by ryanwarsaw** [comment]
@gilrosenthal While setting the value to zero, is an option it would be much handier to have a boolean in the configuration file to disable it all together. Your point about not DoSsing the Overcast Network doesn't make much sense, someone could just spin up a custom instance of the API, remove the caching and DoSs away. 

I also believe that the Overcast Network already has rate limits (requests you can make in a certain time frame) in place, I experimenting with friends and Overcast started rejecting my requests after a couple of dozen. After waiting a few minutes, I was able to resume properly.

**Comment by jengel3** [comment]
Will add this later, for now you can just not connect redis and everything will work just fine, minus the caching.

**Comment by ryanwarsaw** [comment]
@Jake0oo0 Figured I would drop this here, the cache is pretty pointless at this point, since the site already has rate limits that prevent it from receiving too many requests at once.

![](http://i.imgur.com/VpbDAGK.png)

**Comment by jengel3** [comment]
Yes, of course they do @ryanw-se. We also want to be able to return the fastest possible response for our users. Say you want to get the list of tournaments. There would be no need for every query to this endpoint to make a request to Overcast. This list is often the same, and by caching that list we would be able to return a much quicker response. (milliseconds vs several seconds) The same goes for many other endpoints, including categories, and maps. This is specifically why cache times are configurable, and will be able to be enabled or disabled at some point in the future. If you want to increase response times, you can leave our redis caching enabled, or if you need immediate updates you can disable caching.

 It's possible in the future that we could create an endpoint to clear the cache for a certain URL, say if there was a new tournament and that list needed to be updated ASAP. We could also implement a control panel for website admins to manage the caching and see what data is currently being cached. 

tl;dr: Caching on our end allows for faster responses when multiple people, or a single person, request a single endpoint quickly, or an endpoint that does not change often.

**Comment by ryanwarsaw** [comment]
@Jake0oo0 okay so lets say I'm using your API. I depend on it to get me the latest information, from the response there is no way for me to tell if I'm getting fresh information or cached information. There is no way for me to specify in my call to your API if I want a faster response (caching) or not.

Depending on the instance of Sunshine I am depending on to get my information, I'm not sure what the cache times are and if I'm getting the proper information from the proper time frame I want it from. I could want the latest stats for the player Steve, but I could be getting minutes even hours old information instead.

I get what your saying, it might be faster but in its current state it also has a lot of downfalls.

**Comment by jengel3** [comment]
I think most of these are valid concerns, but there is a solution in place for most of them, or will be very soon. 

With caching, yes, it is very important that you know what data you are getting and how recent it is. In this case, we will be providing HTTP headers that specify whether the data is cached or not, and also how long until the data expires. Currently the `X-Response-Type` specifies whether a response is `cached` or `requested`. This allows you to know if you made the API request that retrieved this data, and it is currently accurate. The `X-Cache-Expiry` header will also have a value related to the the time left until the server cache expires, and the next API request will retrieve updated data. 

> There is no way for me to specify in my call to your API if I want a faster response (caching) or not.

This is not intended, and is similar to other APIs, including reddit. However, also similar to reddit, I plan on adding an endpoint such as `/cache/remove` which would be a POST request that includes a URL to remove from the cache. Since we md5 hash the URLs, we can accurately remove a specific URL from the cache, even based on query parameters. 

> I'm not sure what the cache times are 

Regarding this, we could provide an endpoint such as a `HEAD` request to the root URL that returns a JSON body of Sunshine's configuration options, so that you can see how long data is being cached for, or whether it's enabled or not in the first place.

> but I could be getting minutes even hours old information instead.

This is not even possible with our own instance, and caching player data for several hours would make absolutely no sense when it comes to a web API, because you definitely want to have up to date information. On our instance of Sunshine, players are only cached for I believe 30 seconds and this can be easily adjusted based on user feedback. Most endpoints are cached for a very short time, and constantly updated endpoints are cached for minimal times. 

>  it might be faster but in its current state it also has a lot of downfalls.

It is definitely faster, and despite Overcast rate limiting requests, we don't feel that we need to make more requests than we have to. Sunshine is a 3rd party API, and we don't want to abuse the website. Caching is also in place for authentication, which without it, you would be making a request to login every time you make an authenticated API request, and instead we can cached your session token so that it can be reused. Caching has it's upsides, and we can easily solve any downsides that you see, but we need help identifying these issues and solving them, so feel free to contribute changes :smile:

**Comment by ryanwarsaw** [comment]
@Jake0oo0 Thanks for the detailed response, however I still have a few questions. You mentioned having a POST request that can be used to remove data from the cache, so I would need to send that request plus a request to get the new information?

I get how this might be nice for pages like tournaments, or things like teams which don't typically change every day during the off season. But for getting raw player data, people who are querying the information more than once every 30 seconds, probably want to get the latest information.

The players page changes almost constantly, that just makes it even harder to get the information you need by sending a request to remove the data and replace it with new data, every time. 

You also mentioned that your caching authentication requests, that information is only being held in memory, correct? How do you put the pieces together and relate session token x is owned by me making the request?

**Comment by jengel3** [comment]
> more than once every 30 seconds, probably want to get the latest information.

I definitely agree that 30 seconds is too long for this particular endpoint, and we can and will adjust the cache times for several of the endpoints. I added those original times just to finish off caching, and didn't put a lot of thought into each one specifically.

As for the second question, _currently_ when you make a login request to `/players/auth` we create a hash of the email and password, which is cached and mapped to the relevant session token so that we can reuse that for further requests. I don't really like how this works, so we'll likely change it in the near future so that session tokens are encrypted and so that you have an API token, rather than the user/pass hash for making authenticated requests, which will make the API more secure.

A bit of a rushed response this time, but I need to get going, so let me know if you have any more questions.

**Comment by gilrosenthal** [comment]
Yeah this really does not apply to right now so i'm closing it
